### PR TITLE
[Enhancement] Support get partition nams from hms/glue using partition filter

### PR DESCRIPTION
**Pull request by Youngwb**
Fixes #issue
When read hive table, SR need to list all partition names which may get many useless partitition names.
Actually, we could use equals predicate on partition columns to list partition names by partition values.
For example :
 SQL ： 
`select dt,hh,mm from tbl where hh = '12' and mm = '30'; `
 the partition columns are [dt,hh,mm], we could use partition values ["","12","30"]  to get partition names from hms/glue,
 it will return the matched partition names

## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 39 / 116 (33.62%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/hudi/HudiMetadata.java|0|1|00.00%|[91]|
|:large_blue_circle:|com/starrocks/connector/hive/HiveMetadata.java|0|1|00.00%|[89]|
|:large_blue_circle:|com/starrocks/server/MetadataMgr.java|0|10|00.00%|[253, 254, 255, 257, 258, 259, 260, 261, 262, 264]|
|:large_blue_circle:|com/starrocks/connector/hive/HiveMetaClient.java|0|3|00.00%|[195, 196, 197]|
|:large_blue_circle:|com/starrocks/connector/hive/CacheUpdateProcessor.java|0|1|00.00%|[170]|
|:large_blue_circle:|com/starrocks/connector/ConnectorMetadata.java|0|1|00.00%|[99]|
|:large_blue_circle:|com/starrocks/connector/hive/CachingHiveMetastore.java|6|39|15.38%|[210, 211, 212, 213, 214, 215, 216, 217, 221, 222, 224, 225, 235, 243, 244, 246, 248, 383, 384, 386, 484, 485, 532, 533, 542, 543, 550, 551, 553, 576, 577, 591, 592]|
|:large_blue_circle:|com/starrocks/connector/hive/HiveMetastore.java|2|5|40.00%|[95, 96, 97]|
|:large_blue_circle:|com/starrocks/connector/hive/HiveMetastoreOperations.java|1|2|50.00%|[55]|
|:large_blue_circle:|com/starrocks/connector/hive/HivePartitionValue.java|9|16|56.25%|[50, 51, 53, 54, 56, 57, 58]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/OptExternalPartitionPruner.java|21|37|56.76%|[134, 149, 158, 159, 160, 161, 162, 163, 164, 165, 167, 168, 203, 204, 205, 206]|

File: fe/fe-core/src/main/java/com/starrocks/connector/hive/CachingHiveMetastore.java [hunk 1]
```diff
@@ -204,21 +206,53 @@ private List<String> loadAllTableNames(String dbName) {
         return metastore.getAllTableNames(dbName);
     }
 
-    public List<String> getPartitionKeys(String dbName, String tableName) {
+    private List<String> getFilteredPartitionKeys(List<String> partitionNames, List<Optional<String>> partitionValues) {
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/connector/hive/CachingHiveMetastore.java:209]: nit: port this method to PartitionUtils

File: fe/fe-core/src/main/java/com/starrocks/connector/hive/CachingHiveMetastore.java [hunk 2]
```diff
@@ -204,21 +206,53 @@ private List<String> loadAllTableNames(String dbName) {
...
+        List<String> filteredPartitionName = Lists.newArrayList();
+        for (String partitionName : partitionNames) {
+            List<String> values = toPartitionValues(partitionName);
+            int index = 0;
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/connector/hive/CachingHiveMetastore.java:213]: Add a precondition to check size.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/OptExternalPartitionPruner.java [hunk 3]
```diff
@@ -122,6 +125,49 @@ public static LogicalScanOperator prunePartitions(OptimizerContext context, Logi
...
+                collect(Collectors.toList());
+        Map<ColumnRefOperator, ScalarOperator> equalPredicateMap = equalPredicates.stream().
+                collect(Collectors.toMap(rangePredicate -> rangePredicate.getChild(0).cast(),
+                        rangePredicate -> rangePredicate));
+
+        List<Optional<ScalarOperator>> effectivePartitionPredicate = Lists.newArrayList();
+        for (Column partitionColumn : partitionColumns) {
+            ColumnRefOperator partitionColumnRefOperator = operator.getColumnReference(partitionColumn);
+            // only support string type partition column
+            if (partitionColumn.getType().isStringType() && equalPredicateMap.containsKey(partitionColumnRefOperator)) {
```
[the remark this answers is not in the document]
Comment (Youngwb) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/OptExternalPartitionPruner.java:147]: Yes, only string for now

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 39 / 116 (33.62%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/hudi/HudiMetadata.java|0|1|00.00%|[91]|
|:large_blue_circle:|com/starrocks/connector/hive/HiveMetadata.java|0|1|00.00%|[89]|
|:large_blue_circle:|com/starrocks/server/MetadataMgr.java|0|10|00.00%|[251, 252, 253, 255, 256, 257, 258, 259, 260, 262]|
|:large_blue_circle:|com/starrocks/connector/hive/HiveMetaClient.java|0|3|00.00%|[195, 196, 197]|
|:large_blue_circle:|com/starrocks/connector/hive/CacheUpdateProcessor.java|0|1|00.00%|[170]|
|:large_blue_circle:|com/starrocks/connector/ConnectorMetadata.java|0|1|00.00%|[99]|
|:large_blue_circle:|com/starrocks/connector/PartitionUtil.java|0|12|00.00%|[220, 221, 222, 223, 224, 225, 226, 227, 231, 232, 234, 235]|
|:large_blue_circle:|com/starrocks/connector/hive/CachingHiveMetastore.java|6|27|22.22%|[216, 224, 225, 227, 229, 364, 365, 367, 465, 466, 513, 514, 523, 524, 531, 532, 534, 557, 558, 572, 573]|
|:large_blue_circle:|com/starrocks/connector/hive/HiveMetastore.java|2|5|40.00%|[95, 96, 97]|
|:large_blue_circle:|com/starrocks/connector/hive/HiveMetastoreOperations.java|1|2|50.00%|[55]|
|:large_blue_circle:|com/starrocks/connector/hive/HivePartitionValue.java|9|16|56.25%|[50, 51, 53, 54, 56, 57, 58]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/OptExternalPartitionPruner.java|21|37|56.76%|[134, 149, 158, 159, 160, 161, 162, 163, 164, 165, 167, 168, 202, 203, 204, 205]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 39 / 116 (33.62%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/hudi/HudiMetadata.java|0|1|00.00%|[91]|
|:large_blue_circle:|com/starrocks/connector/hive/HiveMetadata.java|0|1|00.00%|[89]|
|:large_blue_circle:|com/starrocks/server/MetadataMgr.java|0|10|00.00%|[251, 252, 253, 255, 256, 257, 258, 259, 260, 262]|
|:large_blue_circle:|com/starrocks/connector/hive/HiveMetaClient.java|0|3|00.00%|[195, 196, 197]|
|:large_blue_circle:|com/starrocks/connector/hive/CacheUpdateProcessor.java|0|1|00.00%|[170]|
|:large_blue_circle:|com/starrocks/connector/ConnectorMetadata.java|0|1|00.00%|[99]|
|:large_blue_circle:|com/starrocks/connector/PartitionUtil.java|0|12|00.00%|[220, 221, 222, 223, 224, 225, 226, 227, 231, 232, 234, 235]|
|:large_blue_circle:|com/starrocks/connector/hive/CachingHiveMetastore.java|6|27|22.22%|[216, 224, 225, 227, 229, 364, 365, 367, 465, 466, 513, 514, 523, 524, 531, 532, 534, 557, 558, 572, 573]|
|:large_blue_circle:|com/starrocks/connector/hive/HiveMetastore.java|2|5|40.00%|[95, 96, 97]|
|:large_blue_circle:|com/starrocks/connector/hive/HiveMetastoreOperations.java|1|2|50.00%|[55]|
|:large_blue_circle:|com/starrocks/connector/hive/HivePartitionValue.java|9|16|56.25%|[50, 51, 53, 54, 56, 57, 58]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/OptExternalPartitionPruner.java|21|37|56.76%|[134, 149, 158, 159, 160, 161, 162, 163, 164, 165, 167, 168, 202, 203, 204, 205]|

**Review by stephen-shelby — APPROVED** [review]

**Review by gengjun-git — APPROVED** [review]

**Comment by Youngwb** [comment]
@Mergifyio  backport branch-3.1

**Comment by XinzhuangL** [comment]
1

**Pull request merged**